File: androidApp/src/main/java/com/cocktailcraft/viewmodel/FavoritesViewModelSKIE.kt
package com.cocktailcraft.viewmodel

import androidx.lifecycle.viewModelScope
import com.cocktailcraft.domain.model.Cocktail
import com.cocktailcraft.viewmodel.SharedFavoritesViewModel
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

/**
 * Android ViewModel wrapper for SharedFavoritesViewModel using SKIE.
 * This replaces the old FavoritesViewModel with a thin wrapper around the shared implementation.
 * 
 * Key SKIE benefits:
 * - StateFlows work directly without conversion
 * - Suspend functions work with Android coroutines
 * - Shared business logic reduces code duplication
 * - Type safety preserved across platforms
 */
class FavoritesViewModelSKIE : BaseViewModel(), KoinComponent {
    
    // Inject the shared ViewModel
    private val sharedViewModel: SharedFavoritesViewModel by inject()
    
    // Expose StateFlows from shared ViewModel (SKIE handles the interop)
    val favorites: StateFlow<List<Cocktail>> = sharedViewModel.favorites
    val favoriteCount: StateFlow<Int> = sharedViewModel.favoriteCount
    
    // Expose loading and error from shared base class
    val loadingState: StateFlow<Boolean> = sharedViewModel.isLoading
    val errorState = sharedViewModel.error
    val errorEventState = sharedViewModel.errorEvent
    
    // Computed properties
    val isEmpty: Boolean
        get() = sharedViewModel.isEmpty
    
    val hasItems: Boolean
        get() = sharedViewModel.hasItems
    
    // MARK: - Public Methods (SKIE converts these to proper suspend functions)
    
    /**
     * Toggle favorite status for a cocktail using SKIE async interop
     */
    fun toggleFavorite(cocktail: Cocktail) {
        viewModelScope.launch {
            sharedViewModel.toggleFavorite(cocktail)
        }
    }
    
    /**
     * Clear all favorites using SKIE async interop
     */
    fun clearAllFavorites() {
        viewModelScope.launch {
            sharedViewModel.clearAllFavorites()
        }
    }
    
    /**
     * Load favorites using SKIE async interop
     */
    fun loadFavorites() {
        viewModelScope.launch {
            sharedViewModel.loadFavorites()
        }
    }
    
    // MARK: - Synchronous Methods (direct delegation)
    
    /**
     * Check if a cocktail is favorite
     */
    fun isFavorite(cocktailId: String): Boolean {
        return sharedViewModel.isFavorite(cocktailId)
    }
    
    /**
     * Get favorites by category
     */
    fun getFavoritesByCategory(category: String): List<Cocktail> {
        return sharedViewModel.getFavoritesByCategory(category)
    }
    
    /**
     * Get favorite categories
     */
    fun getFavoriteCategories(): List<String> {
        return sharedViewModel.getFavoriteCategories()
    }
    
    /**
     * Search favorites by name
     */
    fun searchFavorites(query: String): List<Cocktail> {
        return sharedViewModel.searchFavorites(query)
    }
    
    /**
     * Clear the current error
     */
    fun clearErrorState() {
        sharedViewModel.clearError()
    }
    
    /**
     * Refresh favorites data
     */
    fun refresh() {
        sharedViewModel.refresh()
    }
    
    /**
     * Clean up when ViewModel is cleared
     */
    override fun onCleared() {
        super.onCleared()
        sharedViewModel.onCleared()
    }
}

File: iosApp/CocktailCraft/ViewModels/FavoritesViewModelSKIE.swift
import SwiftUI
import shared

/**
 * iOS ViewModel wrapper for SharedFavoritesViewModel using pure SKIE integration.
 * State/error mirroring and observation-task lifecycle live in
 * SharedViewModelWrapper.
 */
final class FavoritesViewModelSKIE: SharedViewModelWrapper<FavoritesUiState> {

    // Computed properties
    var isEmpty: Bool {
        state.favorites.isEmpty
    }

    var hasItems: Bool {
        !state.favorites.isEmpty
    }

    // Shared ViewModel instance
    private let sharedViewModel: SharedFavoritesViewModel

    init() {
        let viewModel = getSharedKoinHelper().getSharedFavoritesViewModel()
        self.sharedViewModel = viewModel
        super.init(uiState: viewModel.uiState, errorFlow: viewModel.error)
    }

    // No deinit: the base class cancels observation. Wraps a Koin `single`
    // whose coroutine scope must survive any one wrapper — never onCleared().

    // MARK: - Public Methods (using SKIE async/await)

    func loadFavorites() async {
        await run { try await sharedViewModel.loadFavorites() }
    }

    func toggleFavorite(_ cocktail: Cocktail) async {
        await run { try await sharedViewModel.toggleFavorite(cocktail: cocktail) }
    }

    func clearAllFavorites() async {
        await run { try await sharedViewModel.clearAllFavorites() }
    }

    // MARK: - Synchronous Methods

    func isFavorite(_ cocktailId: String) -> Bool {
        return sharedViewModel.isFavorite(cocktailId: cocktailId)
    }

    func getFavoritesByCategory(_ category: String) -> [Cocktail] {
        return sharedViewModel.getFavoritesByCategory(category: category)
    }

    func getFavoriteCategories() -> [String] {
        return sharedViewModel.getFavoriteCategories()
    }

    func searchFavorites(query: String) -> [Cocktail] {
        return sharedViewModel.searchFavorites(query: query)
    }

    func getFavoritesSortedByName() -> [Cocktail] {
        return sharedViewModel.getFavoritesSortedByName()
    }

    func getFavoritesSortedByDate() -> [Cocktail] {
        return sharedViewModel.getFavoritesSortedByDate()
    }

    func getFavoritesSortedByRating() -> [Cocktail] {
        return sharedViewModel.getFavoritesSortedByRating()
    }

    func clearError() {
        sharedViewModel.clearError()
    }

    func refresh() {
        sharedViewModel.refresh()
    }

    // MARK: - Helper Methods for SwiftUI

    func refreshFavorites() async {
        // No artificial delay: the spinner should last exactly as long as the
        // reload does (IO-2).
        await loadFavorites()
    }

    func getCategoryCount(_ category: String) -> Int {
        return getFavoritesByCategory(category).count
    }

    func getTopRatedFavorites(limit: Int = 5) -> [Cocktail] {
        return getFavoritesSortedByRating().prefix(limit).map { $0 }
    }
}
